aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/targeting/endpointProfile/EndpointProfile.kt

```diff
@@ -0,0 +1,332 @@
+            }
+        }
+    }
+
+    /**
+     * Determines if this [EndpointProfile] contains a specific metric.
+     *
+     * @param metricName The name of the metric
+     * @return true if this [EndpointProfile] has a metric with the
+     * specified name, false otherwise
+     */
+    fun hasMetric(metricName: String): Boolean {
+        return metrics.containsKey(metricName)
+    }
+
+    /**
+     * Returns the value of the metric with the specified name.
+     *
+     * @param name The name of the metric to return
+     * @return The metric with the specified name, or null if metric does not
+     * exist
+     */
+    fun getMetric(name: String): Double? {
+        return metrics[name]
+    }
+
+    /**
+     * Adds a metric to this [EndpointProfile] with the specified key. Only
+     * 250 attributes/metrics are allowed to be added to an
+     * [EndpointProfile]. If 250 attribute/metrics already exist on this
+     * [EndpointProfile], the call may be ignored.
+     *
+     * @param name  The name of the metric. The name will be truncated if it
+     * exceeds 50 characters.
+     * @param value The value of the metric.
+     * @return The same [EndpointProfile] instance is returned to allow for
+     * method chaining.
+     */
+    fun withMetric(name: String, value: Double): EndpointProfile {
+        addMetric(name, value)
+        return this
+    }
+
+    /**
+     * Returns a map of all metrics contained within this [EndpointProfile]
+     *
+     * @return a map of all metrics, where the metric names are the keys and the
+     * metric values are the values
+     */
+    val allMetrics: Map<String, Double>
+        get() = Collections.unmodifiableMap(metrics)
+
+    override fun toString(): String {
+        return toJSONObject().toString()
+    }
+
+    fun toJSONObject(): JsonObject {
+        return buildJsonObject {
+            put("ApplicationId", applicationId)
+            put("EndpointId", endpointId)
+            put("ChannelType", channelType.value)
+            put("Address", address)
+            put("Location", Json.encodeToString(EndpointProfileLocation.serializer(), location))
+            put("Demographic", Json.encodeToString(EndpointProfileDemographic.serializer(), demographic))
+            put("EffectiveDate", effectiveDate.millisToIsoDate())
+            put("OptOut", optOut)
+            val attributesJson = buildJsonObject {
+                for ((key, value) in allAttributes) {
+                    try {
+                        put(key, JsonArray(value.map { JsonPrimitive(it) }))
+                    } catch (e: Exception) {
+                        // Do not log e due to potentially sensitive information
+                        LOG.warn("Error serializing attributes.")
+                    }
+                }
+            }
+            // If there are any attributes put then add the attributes to the structure
+            if (attributesJson.isEmpty()) {
```

Comment:
I realized these were backwards while creating unit tests in a forthcoming PR -- I'll correct the logic there as well